rc/core/formatter.kak

```diff
@@ -1,13 +1,19 @@
 decl -docstring "shell command to which the contents of the current buffer is piped" \
     str formatcmd
 
-def format -docstring "Format the contents of the current buffer" %{ eval -draft %{
-    %sh{
-        if [ -n "${kak_opt_formatcmd}" ]; then
-            ## Save the current position of the cursor
-            readonly x=$((kak_cursor_column - 1))
-            readonly y="${kak_cursor_line}"
+# This option holds the selection as it was at the time of calling the `format` command
+# After the buffer has been formatted, the selection will be restored according to
+# the value held in this option
+decl -hidden range-specs formatter_prevsel
```

Comment:
if you do `%|cat<ret>`, the buffer contents is unchanged, but for Kakoune, we did remove the whole content, then insert it again, so any selections will get pushed back to the last char of the buffer (remove whole content makes all selection select the first char, then inserting the buffer in front of that moves all these selections to the end).

The pipe implementation needs to diff the input and output, and then to apply those edits to the buffer. We have the building blocks to generate the diff, what we need is the code that can apply that diff to a buffer.